### DISCUSSION TITLE
How to include gazetteers in the training process of a NER model?

### DISCUSSION THREAD

**Discussion by CarolLuca** [explosion/spaCy#8749] [Help: Coding & Implementations] [answered]
Hello! 

I am asking this question because my two custom NER models have a very low f1 score for the `LOC` enitity:
```python
# trained with ro_core_news_lg vectors

{'ents_p': 0.8711890243902439, 'ents_r': 0.8100637845499645, 'ents_f': 0.8395152405435182, 'ents_per_type': 
{'LEGAL': {'p': 0.9452449567723343, 'r': 0.8179551122194514, 'f': 0.877005347593583}, 
'LOC': {'p': 0.7544910179640718, 'r': 0.7875, 'f': 0.7706422018348622}, 
'ORG': {'p': 0.8229665071770335, 'r': 0.7181628392484343, 'f': 0.7670011148272019}, 
'TIME': {'p': 0.8764940239043825, 'r': 0.9243697478991597, 'f': 0.899795501022495}, 
'PER': {'p': 0.9689922480620154, 'r': 0.9398496240601504, 'f': 0.9541984732824427}}}

AVG = 85.372852771
```
and
```python
# trained with corola.300.20 vectors

{'ents_p': 0.8492840994724944, 'ents_r': 0.7987243090007087, 'ents_f': 0.8232286340394448, 'ents_per_type': 
{'LEGAL': {'p': 0.9058171745152355, 'r': 0.8154613466334164, 'f': 0.8582677165354331}, 
'LOC': {'p': 0.7023809523809523, 'r': 0.7375, 'f': 0.7195121951219512}, 
'ORG': {'p': 0.8211764705882353, 'r': 0.7286012526096033, 'f': 0.7721238938053097}, 
'TIME': {'p': 0.865546218487395, 'r': 0.865546218487395, 'f': 0.865546218487395}, 
'PER': {'p': 0.9407407407407408, 'r': 0.9548872180451128, 'f': 0.9477611940298507}}}

AVG = 83.264224359
```
where, intuitively:
* `LEGAL` refers to legal documents
* `ORG` refers to organizations
* `TIME` refers to time references
* `PER` refers to people's names
* `LOC` refers to places

I do not think very much can be done about `ORG`'s score, but regarding `LOC`'s, I think gazetteers should come in useful. Most of the places mentioned in the data set are either cities, regions from Romania, or big cities, countries from Europe. Creating a list with them (or at least a big part of them) is easy. What I don't know is how to use that list for training my model. I am not sure if it is possible, but the optimal case would be to check if the token appears in the aforementioned list, and, if it does, to mark it as a `LOC`.

Any advice is welcomed!

**Comment by polm** [comment, marked as answer]
spaCy does not support using gazetteers as features for the statistical models the way some NER systems might, but what you can do is use an EntityRuler before your NER component to label names in your gazetteer. If the names are ambiguous that might not work but that sounds like it isn't usually the case in your problem. I would recommend you take a look at the [EntityRuler](https://spacy.io/usage/rule-based-matching#entityruler) docs and see if that helps. 

Note that if you put an EntityRuler before your NER component, the NER component will kind of annotate around the earlier component - it won't overwrite any existing annotations. 

Some other options are using your list of cities to bootstrap training data (weak supervision), or using the EntityRuler as an [annotating component](https://spacy.io/usage/training#annotating-components) during training. Annotating components are a new feature and I'm not sure if we've tested that application of them yet, though it's definitely one we had in mind.

**Reply by CarolLuca** [reply]
Thank you for the clarification! Sorry for answering late, but I tried (and not succeeded) to find some time to implement some of the suggestions. When I will obtain something, I may share the code. Thanks again!
